File: apps/robotlab_x/src/robotlab_x/models/service_config.py
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from typing import List, Any, Optional, Literal, Dict
import time
import json
import yaml
from database.types import DateTimeStr

class ServiceConfig(BaseModel):
    id: Optional[str] = Field(None, description="Primary key - set to service_proxy_id on create", json_schema_extra={"example":"pycv"})
    service_proxy_id: str = Field(..., description="Foreign key - name of the owning service_proxy", json_schema_extra={"example":"pycv"})
    service_meta_id: str = Field(..., description="Denormalized reference to the service catalog entry", json_schema_extra={"example":"opencv@4.8.0"})
    params: Optional[dict] = Field(None, description="Flexible per-instance configuration - structure varies by service type", json_schema_extra={"type":"object","additionalProperties":True,"example":{"resolution":"1920x1080","fps":30}})

# managed
    @classmethod
    def from_yaml(cls, yaml_str: str):
        """Load ServiceConfig from a YAML string."""
        data = yaml.safe_load(yaml_str)
        return cls(**data)

    @classmethod
    def from_yaml_file(cls, file_path: str):
        """Load ServiceConfig from a YAML file."""
        with open(file_path, "r") as file:
            return cls.from_yaml(file.read())

    @classmethod
    def from_json(cls, json_str: str):
        """Load ServiceConfig from a JSON string."""
        data = json.loads(json_str)
        return cls(**data)

    @classmethod
    def from_json_file(cls, file_path: str):
        """Load ServiceConfig from a JSON file."""
        with open(file_path, "r") as file:
            return cls.from_json(file.read())

File: apps/robotlab_x_ui/src/models/ServiceConfig.ts
// managed
// Service Config interface for use across the project
export interface ServiceConfig {
  id?: string;
  service_proxy_id: string;
  service_meta_id: string;
  params?: any;
}

export function createEmptyServiceConfig(): ServiceConfig {
  return {
    id: undefined,
    service_proxy_id: "",
    service_meta_id: "",
    params: undefined,
  };
}

// Helper function to validate ServiceConfig object
export function isServiceConfig(obj: any): obj is ServiceConfig {
  return obj && typeof obj === 'object';
}

// Helper function to clone ServiceConfig object
export function cloneServiceConfig(item: ServiceConfig): ServiceConfig {
  return {
    ...item,
  };
}

// JSON Schema for Fastify route validation and OpenAPI documentation
export const ServiceConfigSchema = {
  "type": "object",
  "properties": {
    "id": {
      "type": "string",
      "description": "Primary key - set to service_proxy_id on create",
      "example": "pycv"
    },
    "service_proxy_id": {
      "type": "string",
      "description": "Foreign key - name of the owning service_proxy",
      "example": "pycv"
    },
    "service_meta_id": {
      "type": "string",
      "description": "Denormalized reference to the service catalog entry",
      "example": "opencv@4.8.0"
    },
    "params": {
      "description": "Flexible per-instance configuration - structure varies by service type",
      "example": {
        "resolution": "1920x1080",
        "fps": 30
      },
      "type": "object",
      "additionalProperties": true
    }
  },
  "required": [
    "service_proxy_id",
    "service_meta_id"
  ]
}
